### ISSUE TITLE
it seems envs were reset twice in venvs

### ISSUE THREAD

**Issue by wujiren** [aai-institute/tianshou#899]
- [ ] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [x] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [ ] new feature request
- [ ] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [x] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, gymnasium as gym, torch, numpy, sys
  print(tianshou.__version__, gym.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  ```

**Comment by wujiren** [comment]
tianshou.__version__: 0.4.11 
gym.__version__: 0.27.1 
torch.__version__: 1.13.1+cu117 
numpy.__version__: 1.23.5 
sys.version: 3.10.9 | packaged by conda-forge | (main, Feb  2 2023, 20:20:04) [GCC 11.3.0] 
sys.platform: linux

i run codes in below, which is mainly test when the env will be reset, it will print env current steps when reset.
```python
from gymnasium import Env
from gymnasium.spaces import Box, Discrete
import numpy as np
from tianshou.env.venvs import SubprocVectorEnv, DummyVectorEnv
from tianshou.policy import PPOPolicy

import gymnasium as gymnasium
import numpy as np
import torch
from tianshou.data import Collector, VectorReplayBuffer
from tianshou.policy import PPOPolicy
from tianshou.utils.net.common import ActorCritic,  Net
from tianshou.utils.net.discrete import Actor, Critic
from tianshou.trainer import OnpolicyTrainer

obs_space = Box(low=0, high=1, shape=(1,))
action_space = Discrete(2)


class MyEnv(Env):
    def __init__(self, my_env_id):
        self.my_env_id = my_env_id
        self.observation_space = obs_space
        self.action_space = action_space
        self.random_episode_length = None
        self.current_step = 0

    def step(self, action):
        self.current_step += 1
        return (
            self.observation_space.sample(),
            0,
            False,
            self.current_step >= self.random_episode_length,
            {},
        )

    def reset(self):
        print(
            f"\nreseting: my env id {self.my_env_id}, current step {self.current_step}, expected episode length {self.random_episode_length}"
        )
        self.current_step = 0
        self.random_episode_length = np.random.randint(5, 10)
        return self.observation_space.sample(), {}

    def seed(self, seed):
        np.random.seed(seed=seed)


def make_env_fn(i):
    return lambda: MyEnv(i)


if __name__ == "__main__":
    device = "cuda" if torch.cuda.is_available() else "cpu"
    train_envs = SubprocVectorEnv([make_env_fn(i) for i in range(1)])
    net = Net(
        obs_space.shape,
        hidden_sizes=[16],
        device=device,
    )
    actor = Actor(net, action_shape=action_space.shape, device=device).to(device)
    critic = Critic(net, device=None).to(device)
    actor_critic = ActorCritic(actor, critic)
    optim = torch.optim.Adam(actor_critic.parameters(), lr=0.001)
    dist = torch.distributions.Categorical
    policy = PPOPolicy(actor, critic, optim, dist)
    train_collector = Collector(
        policy, train_envs, VectorReplayBuffer(100, len(train_envs))
    )
    trainer = OnpolicyTrainer(
        policy=policy,
        train_collector=train_collector,
        test_collector=None,
        max_epoch=2,
        step_per_epoch=20,
        repeat_per_collect=1,
        episode_per_test=0,
        batch_size=32,
        episode_per_collect=1,
    )
    for _ in trainer.run():
        pass

```
codes may too verbose. when i run it, the output likes below:

############################################################################################
reseting: my env id 0, current step 0, expected episode length None
Epoch #1:   0%|                                                                                            | 0/20 [00:00<?, ?it/s]
reseting: my env id 0, current step 7, expected episode length 7

reseting: my env id 0, current step 0, expected episode length 9
Epoch #1:  35%|3| 7/20 [00:00<00:01,  9.76it/s, env_step=7, len=7, loss=0.000, loss/clip=0.000, loss/ent=0.000, loss/vf=0.001, n/e
reseting: my env id 0, current step 7, expected episode length 7

reseting: my env id 0, current step 0, expected episode length 6
Epoch #1:  70%|7| 14/20 [00:00<00:00,  9.76it/s, env_step=14, len=7, loss=0.000, loss/clip=-0.000, loss/ent=0.000, loss/vf=0.001, 
reseting: my env id 0, current step 5, expected episode length 5

reseting: my env id 0, current step 0, expected episode length 6
Epoch #1:  95%|9| 19/20 [00:00<00:00,  9.76it/s, env_step=19, len=5, loss=0.001, loss/clip=-0.000, loss/ent=0.000, loss/vf=0.002, 
reseting: my env id 0, current step 7, expected episode length 7

reseting: my env id 0, current step 0, expected episode length 5
Epoch #1: 26it [00:00,  9.76it/s, env_step=26, len=7, loss=0.001, loss/clip=-0.000, loss/ent=0.000, loss/vf=0.002, n/ep=1, n/st=7,Epoch #1: 26it [00:00, 33.37it/s, env_step=26, len=7, loss=0.001, loss/clip=-0.000, loss/ent=0.000, loss/vf=0.002, n/ep=1, n/st=7, rew=0.00]
Epoch #2:   0%|                                                                                            | 0/20 [00:00<?, ?it/s]
reseting: my env id 0, current step 6, expected episode length 6

reseting: my env id 0, current step 0, expected episode length 5
Epoch #2:  30%|3| 6/20 [00:00<00:00, 387.07it/s, env_step=32, len=6, loss=0.001, loss/clip=-0.000, loss/ent=0.000, loss/vf=0.002, 
reseting: my env id 0, current step 6, expected episode length 6

reseting: my env id 0, current step 0, expected episode length 6
Epoch #2:  60%|6| 12/20 [00:00<00:00, 389.03it/s, env_step=38, len=6, loss=0.001, loss/clip=-0.000, loss/ent=0.000, loss/vf=0.002,
reseting: my env id 0, current step 6, expected episode length 6

reseting: my env id 0, current step 0, expected episode length 8
Epoch #2:  90%|9| 18/20 [00:00<00:00, 388.56it/s, env_step=44, len=6, loss=0.001, loss/clip=-0.000, loss/ent=0.000, loss/vf=0.002,
reseting: my env id 0, current step 5, expected episode length 5

reseting: my env id 0, current step 0, expected episode length 5
Epoch #2: 23it [00:00, 380.47it/s, env_step=49, len=5, loss=0.001, loss/clip=-0.000, loss/ent=0.000, loss/vf=0.002, n/ep=1, n/st=5Epoch #2: 23it [00:00, 379.59it/s, env_step=49, len=5, loss=0.001, loss/clip=-0.000, loss/ent=0.000, loss/vf=0.002, n/ep=1, n/st=5, rew=0.00]
############################################################################################

i think it shows that every time the env was reset twice, and in the second time reset, the current_step is 0.
it will have a similar behavior when use DummyVectorEnv
it's ok because many env reset cost it low, but unfortunately my env reset cost is high, i can not solve the problem once for all.
temporary, when current_step is 0, i will ignore the reset function, can you help me to find a better solution?

**Comment by wujiren** [comment]
my update:
it seems that double reset is reasonable especially in OnpolicyTrainer.
to get a full episode trajectory, we often use param episode_per_collect, let‘s say episode_per_collect =4, and we have a two-env SubprocVectorEnv, that means on average every env runs two times. 
but if the env does not have fixed length, some episode done earlier and some later. let's say now we have collected 3 episode, currently both the two envs are running until one of it finish, and then the data collector stop sampling data, leave another env holds in the middle. 
so to be safe, when start to collect new data, the collector should reset all env to avoid some envs start in the middle. that's why if envs have fixed length will be reset twice.
i think may be it can be optimized but currently it OK.
if i had any mistake about the process, please let me know, and i am going to close the issue.
